### DISCUSSION TITLE
Question about Onboarding renovate.json

### DISCUSSION THREAD

**Discussion by JrobT** [renovatebot/renovate#16801] [Request Help] [answered]
In the docs, under the option `[onboarding](https://docs.renovatebot.com/self-hosted-configuration/#onboarding)` it says:

```
Only set this to false if all three statements are true:
    You've configured Renovate entirely on the bot side (e.g. empty renovate.json in repositories)
```

I didn't realise it was possible to have a shared Organisation level config under the bot repo. In my set up, I have a `renovate-config/default.json` for my run config and then `<repo>/config.json` for my bot config. During onboarding, Renovate then raises an onboarding PR into each repo it's run on which adds a `renovate.json` file which simply extends the `renovate-config/default.json` config.

Is it possible to switch off the onboarding PRs, remove the `renovate.json` from each repo, and house the config in the bot repo as I think the line in the docs suggests?

Ideally, I'd like to reduce the footprint of Renovate across the codebase (i.e. currently 2 repos and a `renovate.json` file in each repo it's run on). I'd like 1 repo for the bot and it's config (or a shared config).

**Comment by PhilipAbed** [comment, marked as answer]
this flag will stop renovate from trying to create an onboarding PR, instead it will look for `renovate.json` file and run 

u can set it in your global config.js or environment variable or cli flag 

although it will not "remove" any `renovate.json` you will have to do that manually, 
renovate.json is mandatory for renovate to run unless you run renovate with another flag [requireConfig](https://docs.renovatebot.com/self-hosted-configuration/#requireconfig)

**Reply by JrobT** [reply]
Okay. How do I set the Renovate config for Renovate to use on each repo via the global config.js/json file?

**Reply by PhilipAbed** [reply]
what do you mean how?  you want an example of a config.js? like mine? there you go: 
config.js:
```
module.exports = {
  token: 'your token',
  platform: 'github',
  includeForks: true,
  onboarding: false,
  repositories: [
	'org/repo',...
  ],
  gitAuthor: 'Self-hosted Renovate Bot <your mail/name>',
};
```

this will run only on the selected `Repositories`, and wont create an onboarding PR, but it might say "requires config" and wont run at all, 
so you will have to 
add `requireConfig: 'optional',`
see https://docs.renovatebot.com/self-hosted-configuration/#requireconfig

or create your own 'reonvate.json' on the repo

**Reply by JrobT** [reply]
I'm not sure if I'm making myself clear but it seems as though the option I'm asking about doesn't exist. Essentially my question is regardless of whether `requireConfig` is optional or not, do I need to have a `renovate.json` file in every repo or can I just define a single `renovate.json` file and tell Renovate to use that file on every repo? In my case, the example config.js you posted above doesn't have any such option like `configFile: './renovate.json'` which I'm imagining would point Renovate to the shared config file to use.

**Reply by PhilipAbed** [reply]
i believe you can use extends: ["github> repo/path/to/renovate"] (renovate.json file) in the config.js

**Reply by viceice** [reply]
you can add **all** config options to the config.js.

**Reply by PhilipAbed** [reply]
also that, u can add anything in `config.js` its the same as `renovate.json` just different syntax and global

**Reply by JrobT** [reply]
thank you.

**Reply by JrobT** [reply]
Hi, just trying this out. I had `onboarding: false` in my `config.json` file, but I still get onboarding PRs raised which try to add the `renovate.json` files to the repos. My config is:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "description": "Default configuration for Renovate",
  "extends": ["local>CMN/renovate-config"],
  "repositories": [
    ....
  ],
  "packageRules": [
    {
      "description": "Group development dependency updates into a single PR once a month",
      "groupName": "Dev dependencies",
      "groupSlug": "all-dev-dependencies",
      "matchDepTypes": ["devDependencies"],
      "schedule": ["before 9am on the first day of the month"]
    }
  ]
}
```

Where `"onboarding": false,` and `"requireConfig": "optional",` are set in the `default.json` file in `renovate-config`.

I was expecting no onboarding PRs to be raised and no `renovate.json` files to be added, and Renovate to just use the configuration in `config.json` as an extension of the config I have in the `renovate-config` repo.

**Reply by rarkins** [reply]
That shouldn't happen. Please run against a single repo with no existing onboarding branch or PR and then post the debug logs to a gist for us to look at

**Reply by JrobT** [reply]
I don't think I can post the entire thing as it has some sensitive information. I'll use a mock repo if that's an issue. This is the part you might be interested in:

```
DEBUG: checkOnboarding() (repository=redacted)
DEBUG: isOnboarded() (repository=redacted)
DEBUG: findFile(renovate.json) (repository=redacted)
DEBUG: findFile(renovate.json5) (repository=redacted)
DEBUG: findFile(.github/renovate.json) (repository=redacted)
DEBUG: findFile(.github/renovate.json5) (repository=redacted)
DEBUG: findFile(.gitlab/renovate.json) (repository=redacted)
DEBUG: findFile(.gitlab/renovate.json5) (repository=redacted)
DEBUG: findFile(.renovaterc) (repository=redacted)
DEBUG: findFile(.renovaterc.json) (repository=redacted)
DEBUG: config file not found (repository=redacted)
DEBUG: findPr(renovate/configure, "Configure Renovate", "!open") (repository=redacted)
DEBUG: getPrList() (repository=redacted)
DEBUG: Retrieved Pull Requests (repository=redacted)
       "length": 43
DEBUG: Renovate did not find a PR for branch #renovate/configure (repository=redacted)
DEBUG: Found no closed onboarding PR (repository=redacted)
DEBUG: Repo is not onboarded (repository=redacted)
DEBUG: getBranchPr(renovate/configure) (repository=redacted)
DEBUG: findPr(renovate/configure, "undefined", "open") (repository=redacted)
DEBUG: getPrList() (repository=redacted)
DEBUG: returning cached PR list (repository=redacted)
DEBUG: Renovate did not find a PR for branch #renovate/configure (repository=redacted)
DEBUG: Onboarding PR does not exist (repository=redacted)
DEBUG: Checking if this org/owner has a default Renovate preset which can be used. (repository=redacted)
DEBUG: Failed to retrieve default.json from repo (repository=redacted)
       "statusCode": 404,
       "url": "https://<bitbucket-endpoint>/rest/api/1.0/projects/ENQ/repos/renovate-config/browse/default.json?limit=20000"
DEBUG: Failed to retrieve renovate.json from repo (repository=redacted)
       "statusCode": 404,
       "url": "https://<bitbucket-endpoint>/rest/api/1.0/projects/ENQ/repos/renovate-config/browse/renovate.json?limit=20000"
DEBUG: Failed to retrieve renovate-config.json from repo (repository=redacted)
       "statusCode": 404,
       "url": "https://<bitbucket-endpoint>/rest/api/1.0/projects/ENQ/repos/.bitbucket-server/browse/renovate-config.json?limit=20000"
DEBUG: No default org/owner preset found, so the default onboarding config will be used instead. Note: do not be concerned with any 404 messages that preceded this. (repository=redacted)
DEBUG: onboarding config (repository=redacted)
       "config": {"$schema": "https://docs.renovatebot.com/renovate-schema.json"}
DEBUG: No renovate config file found (repository=redacted)
DEBUG: migrateAndValidate() (repository=redacted)
DEBUG: No config migration necessary (repository=redacted)
DEBUG: massaged config (repository=redacted)
       "config": {"extends": ["local>CMN/renovate-config"]}
DEBUG: migrated config (repository=redacted)
       "config": {"extends": ["local>CMN/renovate-config"]}
```

**Reply by JrobT** [reply]
settings for `onboarding` and `requireConfig` being in `CMN/renovate-config`

**Reply by rarkins** [reply]
You left off the important config logging part of the log which would let me spot the problem. If you're putting your onboarding config in a preset - ie. it config.json like you previously said - then that's the problem.

**Reply by JrobT** [reply]
Config logs below. I am setting `onboarding` and `requireConfig` in the preset. So they need to move to `config.json`?

```
DEBUG: Using RE2 as regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in /usr/src/app/config.json
DEBUG: File config
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "description": "Default configuration for Renovate",
         "extends": ["local>CMN/renovate-config"],
         "repositories": ["redacted"],
         "packageRules": [
           {
             "description": "Group development dependency updates into a single PR once a month",
             "groupName": "Dev dependencies",
             "groupSlug": "all-dev-dependencies",
             "matchDepTypes": ["devDependencies"],
             "schedule": ["before 9am on the first day of the month"]
           }
         ]
       }
DEBUG: CLI config
       "config": {}
DEBUG: Env config
       "config": {
         "hostRules": [],
         "platform": "bitbucket-server",
         "endpoint": "redacted",
         "username": "redacted",
         "password": "***********",
         "gitAuthor": "redacted <redacted>"
       }
DEBUG: Combined config
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "description": "Default configuration for Renovate",
         "extends": ["local>CMN/renovate-config"],
         "repositories": ["redacted"],
         "packageRules": [
           {
             "description": "Group development dependency updates into a single PR once a month",
             "groupName": "Dev dependencies",
             "groupSlug": "all-dev-dependencies",
             "matchDepTypes": ["devDependencies"],
             "schedule": ["before 9am on the first day of the month"]
           }
         ],
         "hostRules": [],
         "platform": "bitbucket-server",
         "endpoint": "redacted",
         "username": "redacted",
         "password": "***********",
         "gitAuthor": "redacted <redacted>"
       }
```

**Reply by rarkins** [reply]
You need globalExtends instead of extends

**Reply by JrobT** [reply]
Okay. When use that I get a `Missing platform config for local preset.` error, but I assume this is something else. Will look into that.

**Reply by JrobT** [reply]
I'm sorry to keep coming back to you but I'm still struggling with this setup. As the first reply to my question said, "renovate.json is mandatory for renovate to run unless you run renovate with another flag [requireConfig](https://docs.renovatebot.com/self-hosted-configuration/#requireconfig)" - this is not the behaviour I'm experiencing either with a `globalExtends` or without (everything in the same file).

I've done away with the 'globalExtend' config because it was getting in the way and I can look into that later.

When `onboarding: true` I get an onboarding PR that wants to commit a `renovate.json` file to the repo and with `onboarding: false` the repo is skipped because it doesn't have a `renovate.json` file:

```
DEBUG: checkOnboarding() (repository=redacted)
DEBUG: isOnboarded() (repository=redacted)
DEBUG: findFile(renovate.json) (repository=redacted)
DEBUG: findFile(renovate.json5) (repository=redacted)
DEBUG: findFile(.github/renovate.json) (repository=redacted)
DEBUG: findFile(.github/renovate.json5) (repository=redacted)
DEBUG: findFile(.gitlab/renovate.json) (repository=redacted)
DEBUG: findFile(.gitlab/renovate.json5) (repository=redacted)
DEBUG: findFile(.renovaterc) (repository=redacted)
DEBUG: findFile(.renovaterc.json) (repository=redacted)
DEBUG: config file not found (repository=redacted)
 INFO: Repository is disabled - skipping (repository=redacted)
DEBUG: Repository result: disabled-no-config, status: disabled, enabled: false, onboarded: undefined (repository=redacted)
```

This is with `requireConfig: "optional"` in both cases. The `config.json`, set via `RENOVATE_CONFIG_FILE`:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "description": "Default configuration for Renovate per repository",
  "repositories": ["redacted"],
  "onboarding": false,
  "requireConfig": "optional",
  "detectGlobalManagerConfig": true,
  "skipInstalls": false,
  "extends": [
    "config:base",
    "docker:enableMajor",
    ":automergeMinor",
    ":automergePr",
    ":disableDependencyDashboard",
    ":rebaseStalePrs",
    ":semanticCommits",
    ":timezone(Europe/London)"
  ]
}
```

The only thing that I changed between the runs was the `onboarding` flag. What I'm going for is no `renovate.json` in any repo and no onboarding - the PRs are just raised from the config in the above file. My understanding is that running Renovate with the above config should achieve that.

I appreciate all the help you've given me and want to say thank you for being so active in Discussions. I'm happy to close the Discussion if you feel I'm not making myself clear.

**Reply by rarkins** [reply]
onboarding=false and requireConfig=optional should mean that it runs in the repo regardless of whether it has a config

**Reply by JrobT** [reply]
It seems to be this issue: https://github.com/renovatebot/renovate/pull/15640. I am also running `32.62.0`. thanks.

**Comment by simon-liebehenschel** [comment]
@JrobT I think I have a similar problem. Were you able to use the renovate-bot without creating `renovate.json` per repository? I want to define settings only once inside my "bot" repository. Here is my config:
```js
module.exports = {
  // Self-Hosted configuration options
  autodiscover: true,
  autodiscoverFilter: ["/backend/common/"],
  //  baseDir: ".cache",  # This specified in ".gitlab-ci.yml" using environment variables.
  onboarding: false,
  optimizeForDisabled: true,
  persistRepoData: true,
  platform: "gitlab",
  prCommitsPerRunLimit: 20, // The default is no limit, but set some random generous value to stay safe.
  requireConfig: "optional",
  allowedPostUpgradeCommands: ["^poetry version [a-z0-9-]+$"],
  logFile: "renovate.log.json", // Path to log file, must be same as in .gitlab-ci.yml in artifacts

  //  Configuration Options
  hostRules: [
    {
      matchHost: "gitlab.com",
      hostType: "pypi",
      username: "gitlab-ci-token",
      password: process.env.CI_JOB_TOKEN,
    },
  ],
  extends: ["config:base", ":semanticCommits"],
  automerge: true,
  prConcurrentLimit: 5, // The default is no limit, but set some random generous value to stay safe.
  branchNameStrict: true, // I am not sure if we need this, but lets remove special characters to stay safe.
  cloneSubmodules: true,
  dependencyDashboard: true,
  enabledManagers: [
    "dockerfile",
    "docker-compose",
    "gitlabci",
    "poetry",
    "git-submodules",
    "gitlabci-include",
    "pre-commit",
  ],
  //  dnsCache: true, # This is an experimental feature, so maybe enable later.
  enableHttp2: true,
  // keepalive: true,  # This is an experimental feature, so maybe enable later.
  packageRules: [
    {
      matchUpdateTypes: ["minor", "patch"],
      matchCurrentVersion: "!/^0/", // exclude any dependencies which are pre-1.0.0 because those can make breaking changes at any time according to the SemVer spec
      groupName: "all non-major dependencies",
      groupSlug: "all-minor-patch",
    },
  ],
  platformAutomerge: true,
  postUpgradeTasks: {
    commands: ["poetry version patch"], // Source: https://github.com/renovatebot/renovate/issues/16704
    fileFilters: ["pyproject.toml"],
    executionMode: "branch",
  },
  prHourlyLimit: 3, // The default is no limit, but set some random generous value to stay safe.
  reviewers: ["<MASKED>"],
  stabilityDays: 1, // Rarely some version are released by mistake or with a bug. Waiting a little adds some safety.
};
```
I use `renovate/renovate:34.21.4-slim`

My problem is that Renovate returns:
```
 INFO: Repository is disabled - skipping (repository=skynet/backend/common)
DEBUG: Repository result: disabled-no-config, status: disabled, enabled: false, onboarded: undefined (repository=skynet/backend/common)
```

---

**UPD**

I found an unexpected thing. From somewhere Renovate takes `Env config` which I never set, and the renovate merges that "Env config" with my `config.js`... with respect to settings in "Env config"! As a result my settings seems to be overridden...
